azurerm_managed_redis: database persistence

gerrytan opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Adds DB persistence feature to `azurerm_managed_redis` resource and datasource. Persistence cannot be enabled on a geo-replicated db.

<img width="1020" height="732" alt="image" src="https://github.com/user-attachments/assets/38f92dc9-485a-415a-8b83-1a23795eb629" />

Feature doc: https://learn.microsoft.com/azure/redis/how-to-persistence

Example config:

```terraform
resource "azurerm_managed_redis" "test" {
  name     = "my-amr"
  location = "eastus"

  resource_group_name = "my-rg"
  sku_name            = "Balanced_B0"

  default_database {
    persistence_redis_database_backup_frequency = "1h"
  }
}
```

Despite the UX design in portal, we intentionally not opting for single-choice arg (eg: `method = "RedisDatabase"`) to be consistent with the API pattern and anticipate any future API evolution.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- ~~[ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.~~


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_MANAGEDREDIS/527309

<img width="2071" height="1261" alt="image" src="https://github.com/user-attachments/assets/1b380492-a254-44f5-ac00-9c12d52783be" />


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-merging.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_managed_redis` - support for database persistence


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change


## Related Issue(s)

Fixes #30941 


## AI Assistance Disclosure

<!-- 
IMPORTANT!

If you are using any kind of AI/LLM assistance to contribute to the AzureRM provider, this must be disclosed in the pull request.

If this is the case, please check the box below, and include the extent to which AI was used. (e.g. documentation only, code generation, etc.)

If responses to this pull request are/will be generated using AI, disclose this as well.
-->

- [ ] AI Assisted - This contribution was made by, or with the assistance of, AI/LLMs

<!-- extent of AI usage can be described here -->

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

magodo commented on internal/services/managedredis/managed_redis_resource.go:
```diff
@@ -526,6 +556,9 @@ func (r ManagedRedisResource) Update() sdk.ResourceFunc {
 						}
+						if metadata.ResourceData.HasChange("default_database.0.persistence") {
+							dbParams.Properties.Persistence = expandPersistence(state.DefaultDatabase[0].Persistence)
+						}
```
Could it happen if the updated config removes the `default_database.0.persistence`? In this case, dereferencing the 0th index of DefaultDatabase will panic.

gerrytan replied:
Yes default_database can be removed, but it is already handled above. See `case dbLen(old) == 1 && dbLen(new) == 0:` at line 508.

magodo commented on internal/services/managedredis/managed_redis_data_source.go:
```diff
@@ -145,6 +146,24 @@ func (r ManagedRedisDataSource) Attributes() map[string]*pluginsdk.Schema {
+								"backup_frequency": {
+									Type:     pluginsdk.TypeString,
+									Computed: true,
+								},
```
Maybe we shall consider the following schema:

```
"aof": {
  Optional: true,
  ExactlyOneOf: ...,
  {
    "backup_frequency": {},
  },
},

"rdb": {
  Optional: true,
  ExactlyOneOf: ...,
  {
    "backup_frequency": {},
  },
}
```

In this way, we can get rid of the customize diff and additional enum definitions in the validate package.

gerrytan replied:
I'm not sure about that design. If you look at the portal UX, this is designed as a single-choice radio button. In Terraform a block is normally used to group a set of properties.

It should be the swagger / API design that's changed to optimise client use, but I guess it's harder at this stage.

magodo reviewed (changes requested):
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

ziyeqf commented on internal/services/managedredis/managed_redis_data_source.go:
```diff
@@ -255,8 +274,8 @@ func (r ManagedRedisDataSource) Read() sdk.ResourceFunc {
 						GeoReplicationGroupName:         flattenGeoReplicationGroupName(props.GeoReplication),
-						GeoReplicationLinkedDatabaseIds: flattenLinkedDatabases(props.GeoReplication.LinkedDatabases),
 						Module:                          flattenModules(props.Modules),
+						Persistence:                     flattenPersistence(props.Persistence),
```
I rememeber I was commented the data source shall use its own flatten functions instead of sharing the same one with resource. 
But considering this data source is already sharing these functions, it may be acceptable. 
IMO we can keep leave this question to maintainers and if this is the case, the "best practices" document shall be updated

ziyeqf commented on internal/services/managedredis/managed_redis_resource.go:
```diff
@@ -754,3 +804,44 @@ func dbLen(v interface{}) int {
+	case validate.DbPersistenceMethodAOF:
+		res.AofEnabled = pointer.To(true)
+		res.AofFrequency = pointer.ToEnum[databases.AofFrequency](p11Model.BackupFrequency)
+	}
```
can we enable both of them? Even if the Portal shows it as single option, what's the API behavior?

ziyeqf commented:
Thanks @gerrytan for the PR, left some comments above

gerrytan commented on internal/services/managedredis/managed_redis_resource.go:
```diff
@@ -195,6 +203,52 @@ func (r ManagedRedisResource) Arguments() map[string]*pluginsdk.Schema {
+										"default_database.0.persistence.0.redis_database_enabled",
+										"default_database.0.persistence.0.append_only_file_enabled",
+									},
+									ValidateFunc: validate.BoolIsTrue,
```
Because this config does not make any sense:

```
default_database {
    persistence {
      redis_database_enabled          = false
      redis_database_backup_frequency = "1h"
    }
}
```

Instead of specifying that, user should remove the entire `persistence` block instead

jackofallops commented on internal/services/managedredis/managed_redis_data_source_test.go:
```diff
@@ -51,6 +52,21 @@ func TestAccManagedRedisDataSource_basic(t *testing.T) {
 	})
 }
 
+func TestAccManagedRedisDataSource_dbPersistence(t *testing.T) {
```
Data sources don't typically have more than 1 test (complete), can we just extend the existing test to accommodate the new properties?

gerrytan commented:
>Typically we do split the concerns. Data Sources and their respective resources regularly differ in their schemas. For instance, is there any actual user value in exposing the new persistence block in the Data Source. I.e. can the values be meaningfully used by users elsewhere?

@jackofallops I just add them simply based on consistency between resource and data source. Happy to take this off if you think it's unnecessary.

gerrytan commented:
@WodansSon @jackofallops great minds think alike! 🫡. I have refactored the schema to be flat two properties. Please checkout the latest example in the PR desc. 

Acctest for latest commit is also in progress.

jackofallops commented on website/docs/r/managed_redis.html.markdown:
```diff
@@ -154,10 +154,14 @@ A `default_database` block supports the following:
 
 * `module` - (Optional) A `module` block as defined below. Refer to [the modules documentation](https://learn.microsoft.com/azure/redis/redis-modules) to learn more.
 
+* `persistence_append_only_file_backup_frequency_in_seconds` - (Optional) The frequency of Append Only File (AOF) backups. Possible value is `1`. Providing this value implies AOF persistence method is enabled. Conflicts with `persistence_redis_database_backup_frequency_in_hours`, only one persistence method is allowed. Conflicts with `geo_replication_group_name`, persistence can only be enabled o ...
```
The enum exposed is `1s`, not `1`?

jackofallops replied:
The "exception" here is that the property was unfortunately misnamed, since the values are `Durations` not numerical values. I'll check the docs PR asap to make sure that's clear, but thanks for following that up 👍

gerrytan commented:
Hello! Thanks for reviewing @jackofallops . I have reverted the property back to enums and killed the string-integer mapping.

I have also created this PR to improve reference-naming.md to help future contributors: https://github.com/hashicorp/terraform-provider-azurerm/pull/31054

Please have another look 🙏

jackofallops commented on internal/services/managedredis/managed_redis_resource.go:
```diff
@@ -754,3 +791,41 @@ func dbLen(v interface{}) int {
+	if pointer.From(input.AofEnabled) {
+		return pointer.FromEnum(input.AofFrequency)
+	}
+	return ""
```
nit
```suggestion
	if input == nil {
		return ""
	}
	
	if pointer.From(input.AofEnabled) {
		return pointer.FromEnum(input.AofFrequency)
	}
	
	return ""
```

gerrytan commented on internal/services/managedredis/managed_redis_resource.go:
```diff
@@ -754,3 +791,41 @@ func dbLen(v interface{}) int {
+			RdbEnabled:   pointer.To(true),
+			RdbFrequency: pointer.ToEnum[databases.RdbFrequency](rdbBackupFreq),
+		}
+	}
```
Fixed in 8d2a7cdf2b

gerrytan commented on internal/services/managedredis/validate/managed_redis_database_persistence.go:
```diff
@@ -0,0 +1,20 @@
+			res = append(res, freq)
+		}
+	}
+	return res
```
Fixed in 8d2a7cdf2b

gerrytan commented:
@jackofallops I have addressed all feedbacks, fixed out-of-sync docs, and resolved conflict from main. Acctest is in progress. Can you please take another look 🙏

jackofallops commented on internal/services/managedredis/managed_redis_resource_test.go:
```diff
@@ -183,6 +240,8 @@ resource "azurerm_managed_redis" "test" {
   location = "%[2]s"
   sku_name = "Balanced_B0"
+
+  default_database {}
```
Understood, but the purpose of a basic test is to cover `Required` properties only, so this should be removed and other configs added to satisfy other purposes. Can you update for this?

gerrytan replied:
acctest run in progress: https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_MANAGEDREDIS/527309

gerrytan commented:
Thank you @jackofallops 🙌
